ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/acceptance/infrastructure/LedgerSession.scala

```diff
@@ -0,0 +1,274 @@
+    for {
+      id <- ledgerId()
+      end <- ledgerEnd()
+      responses <- FiniteStreamObserver[GetTransactionsResponse](
+        transactionService
+          .getTransactions(
+            new GetTransactionsRequest(
+              ledgerId = id,
+              begin = Some(begin),
+              end = Some(end),
+              filter = Some(transactionFilter(party, templateIds)),
+              verbose = true
+            ),
+            _
+          ))
+    } yield responses.flatMap(_.transactions)
+
+  private[infrastructure] def clock()(implicit context: ExecutionContext): Future[Clock] =
+    for (id <- ledgerId()) yield new LedgerClock(id, timeService)
+
+  private[infrastructure] def create(
+      party: String,
+      templateId: Identifier,
+      args: (String, Value.Sum)*)(implicit context: ExecutionContext): Future[String] =
+    for {
+      id <- ledgerId()
+      clock <- clock()
+      now = clock.instant()
+      inFiveSeconds = now.plusSeconds(5)
+      tx <- commandService.submitAndWaitForTransaction(
+        new SubmitAndWaitRequest(
+          Some(new Commands(
+            ledgerId = id,
+            applicationId = UUID.randomUUID().toString,
+            commandId = UUID.randomUUID().toString,
+            party = party,
+            ledgerEffectiveTime = Some(new Timestamp(now.getEpochSecond, now.getNano)),
+            maximumRecordTime =
+              Some(new Timestamp(inFiveSeconds.getEpochSecond, inFiveSeconds.getNano)),
+            commands = Seq(
+              new Command(
+                Create(
+                  new CreateCommand(
+                    Some(templateId),
+                    Some(new Record(
+                      fields = args.map {
+                        case (label, value) =>
+                          new RecordField(label, Some(new Value(value)))
+                      }
+                    ))
+                  )
+                )
+              )
+            )
+          ))))
+    } yield tx.transaction.get.events.collect { case Event(Created(e)) => e.contractId }.head
+
+  private[infrastructure] def exercise(
+      party: String,
+      templateId: Identifier,
+      contractId: String,
+      choice: String,
+      args: (String, Value.Sum)*
+  )(implicit context: ExecutionContext): Future[Unit] =
+    for {
+      id <- ledgerId()
+      clock <- clock()
+      now = clock.instant()
+      inFiveSeconds = now.plusSeconds(5)
+      _ <- commandService.submitAndWait(
+        new SubmitAndWaitRequest(
+          Some(
+            new Commands(
+              ledgerId = id,
+              applicationId = UUID.randomUUID().toString,
+              commandId = UUID.randomUUID().toString,
+              party = party,
+              ledgerEffectiveTime = Some(new Timestamp(now.getEpochSecond, now.getNano)),
+              maximumRecordTime =
+                Some(new Timestamp(inFiveSeconds.getEpochSecond, inFiveSeconds.getNano)),
+              commands = Seq(
```

Comment:
Considering that `commands` is the only difference between `create` and `exercise`, maybe we can deduplicate the fluff around it.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/5094794b8fe366b4f1ba082f1fb87a8190672797

This commit also adds helpers specific for test DAML code.